### PR TITLE
Implement AUTO_REFRESH_MAX_TTL to limit total token lifetime when AUTO_REFRESH = True.

### PR DESCRIPTION
**Pull request by christian-oudard**
Refs https://github.com/jazzband/django-rest-knox/issues/311

### CODE REVIEW

**Review by johnraz — CHANGES_REQUESTED** [review]
Thanks a lot for the contribution, I have a few comments 😉

File: knox/auth.py
Comment (christian-oudard) [reply knox/auth.py:80]: I guess I don't understand the library enough to get what needs to be different here. This seems to me like an elegant solution to not extending the maximum time since creation date. Is there a different exception or http code or something you want to return to the user instead?

Comment (johnraz) [reply knox/auth.py]: I would like the error message to clearly state that the token couldn’t be auto refreshed due to the max ttl being reached yes.

While this code would work, it would be quite hard to understand what happened.

Remember the Zen of Python: explicit is better than implicit 😉

Comment (christian-oudard) [reply knox/auth.py]: Okay, but the current design deletes token objects once they expire. The error message does not currently distinguish between tokens that never existed and ones that once existed then expired. According to the docstring of `authenticate_credentials`: "Tokens that have expired will be deleted and skipped", using `_cleanup_token`.

So I think it's on you to provide a mechanism to explicitly distinguish these things.

Comment (johnraz) [reply knox/auth.py]: This is not really true.

Nothing prevents us from raising something like ˋAutoRefreshMaxTTLExpired` here and handling the exception in `authenticate_credentialsˋ in order to return a specific error message  with ˋ exceptions.AuthenticationFailed`.

Marking the token has expired and letting it continue its life until it reaches another step of the flow feels wrong to me, really.

Comment (johnraz) [reply knox/auth.py]: Reading the code again I realize we actually don’t want an error here as the authentication passed … Then I guess adding a log message stating that the max ttl was reached would be better already.

And maybe for the longer game we should review the way this is designed: decide if a token should be refreshed at the time we identify it is actually expired - this would give us more room for raising appropriate errors but this is definitely outside of the scope of this PR

Comment (christian-oudard) [reply knox/auth.py]: Totally fine adding a log message, but there is currently no logging whatsoever in the project, so I would wait for you or one of the other main developers to add that before trying to do it myself. My client is already using our fork of this in its current state, so I'm going to stop working on this PR for now. Please feel free to do anything you need to to merge this into the main codebase.

Comment (johnraz) [reply knox/auth.py]: AFAIK logging configuration is handled by Django.

To add a log line here you can just follow the usual way:

`logger = logging.getLogger(__name__)`

and then use the logger accordingly.

File: tests/tests.py
Comment (johnraz) [reply tests/tests.py:349]: Maybe I wasn’t clear in my previous comment but can you add the call that actually fails due to the expired token ?

Comment (christian-oudard) [reply tests/tests.py:349]: Added an http request to the test and showed that it fails.

File: knox/auth.py
Comment (christian-oudard) [reply knox/auth.py]: Added a log message when the expiry is capped by AUTO_REFRESH_MAX_TTL.

**Review by johnraz — APPROVED** [review]
Looks good to me, can you look into why you have a pre-commit related commit in your PR?

After that is solved we should be good to merge 😁

@giovannicimolin maybe you want to give this a check too?

**Comment by christian-oudard** [comment]
Any predictions when you'll be able to merge this?

**Comment by johnraz** [comment]
When @giovannicimolin will have free time to review so no, there is no specific timeline

**Comment by eranrund** [comment]
Looking forward to seeing this merged :) Thank you!

**Review by giovannicimolin — APPROVED** [review]
:+1: 

Nice addition! Looks good to go 

@johnraz Thanks for the heavy lifting here!

**Pull request merged**
